### ISSUE TITLE
Generator memory leak bugs?

### ISSUE THREAD

**Issue by BIGBALLON** [keras-team/keras#8668]
I used 4 GPUs to train DenseNet(layer=190, k = 40).

- tensorflow-gpu (1.4.0)
- Keras (2.1.2)

It seems that during training via model.fit_generator(), 
the system memory increase gradually until program crash.

The training have 250 epochs.
At the first epoch, Mem used [ 4.62/15.6GB ], 
after about 50 epochs, [ 15.5GB/15.6GB ], 
then program crash.

**Comment by tRosenflanz** [comment]
https://github.com/fchollet/keras/pull/8666 should have fixed it. Please pull the latest version and retest.

**Comment by apedevmicrosoft** [comment]
I am having the same issue. I pulled your latest code with the "closing" statement in it and I am still having the issue.

**Comment by BIGBALLON** [comment]
@apedevmicrosoft  I clone the latest code & setup, still having the issue.

**Comment by apedevmicrosoft** [comment]
@tRosenflanz  I have a feeling that there is still something going on with the threads. If I play with `workers` and `max_queue_size` parameters I can get into stable and non stable states. It feels like there is something that is not able to consume the data from the queue fast enough and the memory fills up. I looked at the `OrderedEnqueuer` and I dont see anything obvious. I have scenarios where the moment I increase `workers` by 1 the memory starts filling up but if I keep it below a certain threshold it stays consistent. GPU utilization ranges from 70-85% depending on the number of `workers`. Any ideas?

**Comment by tRosenflanz** [comment]
@apedevmicrosoft I am just a user myself, there was an identical bug that got fixed but that update. @Dref360 I think this is up your valley

**Comment by Dref360** [comment]
Any of you have a reproducible example? I'll be able to look more into it around X-mas. (Finals and stuff)

**Comment by Dref360** [comment]
Some questions to guide me : 
1. OS and python version
2. `multiprocessing=True` or `False`
3. Value of `workers` and `queue_size`.
4. Spec of your PC (RAM, Video card)

**Comment by apedevmicrosoft** [comment]
1. OS and python version
Win 10, Conda 4.3.27, Python 3.5.4
2. multiprocessing=True or False
False
3. Value of workers and queue_size.
leaks: workers=6, max_queue_size=512
4. Spec of your PC (RAM, Video card)
RAM: 64gb GPU: Titan XP 12gb

I will create a sample in a bit.

**Comment by Dref360** [comment]
If it's a Windows only issue, I won't be able to work on it. Since there is no fork on Windows, Python does a lot of magic to make it work. Maybe that's the issue?

**Comment by apedevmicrosoft** [comment]
I dont think its a Windows issue. @tRosenflanz  is using "multiprocessing=True" which follows a different code path.

**Comment by Dref360** [comment]
Are you using a generator or a Sequence?
Also, using more than one worker with `multiprocessing=False` has no benefit because of the GIL

**Comment by apedevmicrosoft** [comment]
I am using `ImageDataGenerator.flow_from_directory` which takes the code down the Sequence path internally (you would imagine it would go down the generator path, but it does not when I debug it). 

As for the second part, I do see a significant gain when using `multiprocessing=False` and adding multiple workers as it does seem to create more threads to pre-process the data.

**Comment by Dref360** [comment]
Yeah, Sequence work with multiprocessing=False, 
I'm currently unable to reproduce your issue, maybe you could share a gist?

**Comment by apedevmicrosoft** [comment]
Ill work on getting you a sample. How large is your sample dataset? Its very prominent when using train and valid generator on a relatively large dataset, such as Imagenet.

**Comment by Dref360** [comment]
100k 300x300 generated images

**Comment by apedevmicrosoft** [comment]
1. NumTrainingImages: 1,281,167
2. NumValImages: 500,000
3. InputShape: (240,240)
4. BatchSize: 64

**Comment by apedevmicrosoft** [comment]
Seems like people are seeing the same issue in this thread: [https://github.com/keras-team/keras/issues/5835#issuecomment-353629656](https://github.com/keras-team/keras/issues/5835#issuecomment-353629656)

**Comment by Dref360** [comment]
With your values, you queue can go up to 40 GB of RAM. 
2 queues (train/val) * 512 queue_size * (64 * 240 * 240 * 3) * 4 byte.

Are you seeing any zombie threads again? Or it's just that the queue is filling up?

**Comment by alexhernandezgarcia** [comment]
Thank you @apedevmicrosoft for pointing me here from #5835. I am having a similar issue. This are the system details:

- OS: Ubuntu 16.04.2 LTS
- RAM: 128 GB
- GPU: NVIDIA GeForce GTX 1080 Ti 11GB
- Python 2.7.12; Keras 2.1.2; Tensorflow 1.2.0

The data set is ImageNet ILSVRC 2012, with the following characteristics:

- Number of training images: 1,281,167
- Number of validation images: 50,000
- Image shape: [150, 200, 3]
- Encoding: Float 32

What I have observed so far: I try to train with `fit_generator`. Before starting training, the RAM memory gets filled slowly up to the 128 GB available, then the swap memory and at some point (when Swp is about 100 GB) I get a `MemoryError`.

This happens with the default values of `max_queue_size=10, workers=1, use_multiprocessing=False`. But I have also tried with `max_queue_size=1, workers=10, use_multiprocessing=True` and others with identical result. I have also found the batch size irrelevant (have tried with 16 instead).

Finally, the only way I have managed to enable the training with ImageNet is by reducing the number of training images to 10 % (128,116). With 50 % (640583) I get the memory error as well. 

I am not sure about all the details of how `fit_generator` works, but my intuition is that it is trying to allocate on RAM at least as many images as the amount of available images for training, which does not make sense. If I am not wrong, this was working with a previous version of Keras, but I do not remember which and I haven't tried again.

I am happy to help find out what's going on if you have some time too.

**Comment by Dref360** [comment]
Are you too using ImageDataGenerator? If so, with flow or flow_from_directory?

**Comment by alexhernandezgarcia** [comment]
Yes, I am using ImageDataGenerator with `flow`.

**Comment by Dref360** [comment]
Could you give me a quick snippet of your generator?
`flow` takes 2 numpy array so you call `flow` on each batch?

**Comment by alexhernandezgarcia** [comment]
```
# Create batch generators
batch_gen_tr = ImageDataGenerator(width_shift_range=0.1,
                                  height_shift_range=0.1,
                                  horizontal_flip=True)
batch_gen_val = ImageDataGenerator()
```

```
# Train model
model.fit_generator(generator=batch_gen_tr.flow(images_tr, labels_tr),
                    steps_per_epoch=20000,
                    epochs=25,
                    validation_data=batch_gen_val.flow(images_val, labels_val),
                    validation_steps=780)
```

`images_tr`, `images_val`, `labels_tr` and `labels_val` are `dask` arrays in my case.

**Comment by Dref360** [comment]
So the shape of `image_tr` is ~ [1281167,150, 200, 3] ?

**Comment by alexhernandezgarcia** [comment]
Yes, exactly.

**Comment by Dref360** [comment]
Then your memory is already filled up with the images, it's not a bug. You load 1 million images in memory.
You need to use a generator to create your samples

**Comment by alexhernandezgarcia** [comment]
Sorry, I forgot the `batch_size` argument in the `flow` function. It should be:

```
model.fit_generator(generator=batch_gen_tr.flow(images_tr, labels_tr, batch_size=64),
                    steps_per_epoch=20000,
                    epochs=25)
```

So, if I am not wrong, `fit_generator` should be able to retrieve the data batch by batch, without filling up the memory. 

`images_tr` can be [1281167, 150,  200, 3] because it is a `dask` array, so it is loaded on hard disk, not RAM.

**Comment by Dref360** [comment]
ImageDatagenerator cast `images_tr` to an array. [here](https://github.com/keras-team/keras/blob/master/keras/preprocessing/image.py#L881)
Feel free to submit a PR if you want to avoid this behavior. Or use a true generator.

**Comment by alexhernandezgarcia** [comment]
Oh I see, thanks! I think it would be better if Keras was able to handle large dask arrays. The API would be more transparent. Then, is it not possible to use `fit_generator` with large data sets? What do you mean by using a "true generator", @Dref360?

**Comment by Dref360** [comment]
Something like : 
```python
img_paths = ... # Path to all the images
datagen = ImageDataGenerator(...)
def your_gen(paths):
    while True:
         for X_batch, y_batch in get_batches(paths):
             X_batch = [datagen.random_transform(x) for x in X_batch]
             # additional preprocessing
             yield X_batch, y_batch

model.fit_generator(your_gen(img_path),...)
```

**Comment by apedevmicrosoft** [comment]
@Dref360 

> Are you seeing any zombie threads again? Or it's just that the queue is filling up?

I am actually not exactly sure where the problem is. Could you suggest a way to debug this?

**Comment by Ricocotam** [comment]
Hi, i was looking for this exact issue and i found this one #8677. Are you guys using a call back ? Because it seems threads are started again for both validation and after callbacks. And I do have zombies.

**Comment by apedevmicrosoft** [comment]
These are the callbacks that I am using:

```
string_start_time = datetime.fromtimestamp(time.time()).strftime('%Y-%m-%dT%H-%M-%SZ')

lr_reducer      = ReduceLROnPlateau(monitor='val_loss', factor=np.sqrt(0.1),
                                    cooldown=0, patience=20, min_lr=0.5e-6)
lr_logger = LearningRateLogger(lr_file)

early_stopper   = EarlyStopping(monitor='val_acc', min_delta=0.0001, patience=50)
model_checkpoint= ModelCheckpoint(model_file, monitor="val_acc", save_best_only=True,
                                  save_weights_only=False)

callbacks=[lr_reducer, lr_logger, early_stopper, model_checkpoint]

if K.backend() == 'tensorflow':
    tensorboard = TensorBoard(log_dir='logs/{}'.format(string_start_time)) #, histogram_freq=5, batch_size=batch_size)
    callbacks.append(tensorboard)
```

**Comment by Ricocotam** [comment]
I tried to run the same model i had (xception on transfert learning) but without call backs as it is suggested on #8677, I still get a crash. I read on #8946 that it's been corrected on master (#8666) so I'll get a new try.

**Comment by mohapatras** [comment]
Any updates on this ?

**Comment by Ricocotam** [comment]
Yes, I just update keras and it works well.

**Comment by Haseeb92** [comment]
Working with Keras 2.1.5, Google colab, still crashes with a batch size of 100 article images, and 3 layers of LSTM. When I downgrade to 2 layers, it starts working again. model.fit_generator seems to explode in memory.

**Comment by Dref360** [comment]
If it works with 2 layers and not 3, it's your model that is too big. Not fit_generator related.

**Comment by JSAustin** [comment]
I am having a similar issue, where a memory leak occurs during training with fit_generator.  Here is my network/training scheme.  I am using all up-to-date keras, numpy, etc... with 64 GB RAM, GTX 1080.  If I use fit instead of fit_generator, it flies through and uses only a fraction of my memory.  fit_generator fills up the memory.  Changing use_multiprocessing to False or changing workers doesn't seem to do anything.

My images are 160x320.  The X_train vector has about 30k images in it.  X_valid is about 10k images.

# Create Keras Model
conv_layer_depth = (24,36,48,60,72,84)
dense_depth = (150,75)
dropout_rate = 0.5
learning_rate=0.002
num_epoch = 30
batch_size = 64

model = Sequential()
model.add(BatchNormalization(mode=0, axis=3, input_shape=X_train.shape[1:]))
model.add(Conv2D(conv_layer_depth[0], (5,5), strides=(1,1), padding='valid'))
model.add(Activation('relu'))
model.add(Conv2D(conv_layer_depth[1], (3,3), strides=(2,2), padding='valid'))
model.add(Activation('relu'))
model.add(MaxPooling2D(pool_size=(2,3), strides=(2,2), padding='valid'))
model.add(Dropout(dropout_rate*0.1))

model.add(Conv2D(conv_layer_depth[2], (3,3), strides=(1,1), padding='valid'))
model.add(Activation('relu'))
model.add(Conv2D(conv_layer_depth[3], (3,3), strides=(2,3), padding='valid'))
model.add(Activation('relu'))
model.add(Dropout(dropout_rate*0.2))

model.add(Conv2D(conv_layer_depth[4], (3,3), strides=(2,2), padding='same'))
model.add(Activation('relu'))
model.add(Dropout(dropout_rate*0.3))

model.add(Conv2D(conv_layer_depth[5], (3,3), strides=(2,2), padding='same'))
model.add(Activation('relu'))
model.add(Dropout(dropout_rate*0.4))

model.add(Conv2D(conv_layer_depth[5], (3,3), strides=(2,2), padding='same'))
model.add(Activation('relu'))
model.add(Dropout(dropout_rate*0.4))

model.add(Flatten())

model.add(Dense(dense_depth[0]))
model.add(Activation('relu'))
model.add(Dropout(dropout_rate))
model.add(Dense(dense_depth[1]))
model.add(Activation('relu'))
model.add(Dropout(dropout_rate))
model.add(Dense(1))
model.add(Activation('linear'))

opt = Adam(lr=learning_rate, decay=1e-6)
model.compile(loss='mse',
             optimizer=opt)

datagen = ImageDataGenerator(
        rotation_range=15, 
        width_shift_range=0.1, 
        height_shift_range=0.1) 
datagen.fit(X_train)

model.fit_generator(datagen.flow(X_train, y_train,
                                 batch_size=batch_size),
                                 epochs=num_epoch,
                                 validation_data=(X_valid, y_valid),
                                 workers=2,
                                 use_multiprocessing=True)

**Comment by adt58780** [comment]
This issue might be caused by Tensorflow, rather than Keras.
I was working on Keras 2.2.4 with Tensorflow 1.14.0(CPU) backend, and I had the same issue. Then I downgraded the Tensorflow to 1.13.1 and I found out no memory is leaking anymore. I haven't changed my python script or Keras version.

**Comment by ale152** [comment]
I have a similar problem to the one discussed in this issue. Training the network increases the memory batch after batch, until the program gets killed because of memory exceeding the limits.

I'm using Keras 2.2.4 and 1.12.0. I currently have 128GB of RAM and my dataset occupies 20 GB (compressed). I load the compressed dataset in RAM and then use a Sequence generator to uncompress batches of data to feed the network. Each of my samples is 100x100x100 pixels stored as float32, I'm using a batch_size of 64, queue_size of 5 and 27 workers with multiprocessing=True. In theory, I should have a total of 100*100*100 *4 *64 *5 *27 =~ 35 GB. However, when I run my script, it gets killed by the queuing system because of excessive memory usage:
```
slurmstepd: error: Job XXXX exceeded memory limit (1192359452 > 131072000), being killed
```
I've even tried to use a max_queue_size as little as 2, and the process is still exceeding the maximum memory. To make things even more difficult to understand, sometimes, completely by chance, the process is executed properly (even with a max_queue_size of 30!).

I'm really going crazy with this issue...

**Comment by DeeperDeeper** [comment]
Hi @Dref360,
I have the same problem. I am using Tensorflow 1.14 and keras 2.2.4 on ubuntu, GPU 1080 GTX cuda 9 and cudnn 7.
I am using keras fit_generator to train the model. The generator is done by Tensorpack (dataflow) and it is tested correctly for 2000 iterations.
Batch_size = 4, image_size =512x512x3, dataset size =2000 images
fit_generator with default parameters.
I see that the RAM is keep increasing while the model is training, from 50% to 75% for only 70 epochs!

**Comment by AlexVaith** [comment]
I am facing the same issue with numpy arrays as an input. The memory just increases with every batch and is not cleared. Is there a solution to the problem by now? Or do we all understand the concept of the fit_generator wrong?
From what I thought, the fit_generator is supposed to load a new sample for every batch. But after this data is processed by the model (forward- and backward propagation), is hould be released from memory: Is this correct?

**Comment by Shashank9830** [comment]
Just wanted to mention that I'm facing the exact same problem while using fit_generator. The memory inside RAM keeps on increasing and finally I get a ResourceExhaustedError at some point. Looks like memory from the previous batch is not released or released very slowly leading to this. I have tried creating very small batches of like 8 images. So, I get a numpy array that's not too large. Still after some number of epochs memory is released very slowly and finally it just crashes. I'm running it on Kaggle kernel with 16 GB of RAM.

**Comment by yaochenzhu** [comment]
I am using tf 1.15 to train an encoder-decoder-like model, and running into the same problem when I call the .predict method of the encoder every epoch. (GPU memory explodes after around 100 epoches). Simply changing the .predict to .predict_on_batch with nothing else altered seems fix my problem, and there seems no demage to the performance.

**Comment by mokundong** [comment]
I found this problem too with ubuntu18.04 and anaconda,but when i use the python3 which inside in ubuntu,bug fixed.

**Comment by atotev** [comment]
https://fantashit.com/linearly-increasing-memory-with-use-multiprocessing-and-keras-sequence/#comment-254237
